File: js/src/library/apply.js
/**
 * Apply all changes immutably to the source object.
 *
 * @param {[key: string, value: any][]} changes The changes to be applied
 */
export function applyObject(source, changes) {
  let result = source;
  let unchanged = true;

  changes.forEach((entry) => {
    const [key, value] = entry;

    if (entry.length === 1) {
      if (!Object.hasOwn(result, key)) return;
      if (unchanged) {
        let _;
        ({ [key]: _, ...result } = result);
        unchanged = false;
      } else {
        delete result[key];
      }
      return;
    }

    if (Object.hasOwn(result, key) && result[key] === value) return;
    if (unchanged) {
      result = { ...result, [key]: value };
      unchanged = false;
    } else {
      result[key] = value;
    }
  });

  return result;
}

/**
 * Apply all changes immutably to the source array.
 * Indices can be negative to be applied from the end of the array.
 *
 * @param {[index: string, value: any][]} changes The changes to be applied
 * @param filler Value to be used as padding
 */
export function applyArray(source, changes, filler = null) {
  let result = source;
  let unchanged = true;

  changes.forEach((entry) => {
    const [index, value] = entry;
    let i = index >= 0 ? index : result.length + index;

    if (i >= 0 && i < result.length && result[i] === value) return;
    if (index >= 0) {
      const suf = i + 1 - result.length;
      if (suf > 0) {
        result = [...result, ...new Array(suf).fill(filler)];
        unchanged = false;
      }
    } else {
      const pre = -i;
      if (pre > 0) {
        result = [...new Array(pre).fill(filler), ...result];
        unchanged = false;
        i = 0;
      }
    }
    if (unchanged) {
      result = [...result];
      unchanged = false;
    }
    result[i] = value;
  });

  return result;
}

/**
 * Create all possible combinations immutably based on the specified source array.
 * If the specified `combinations` array is empty, the resulting array contains a single empty array.
 * This function has essentially the same base functionality as the `mux` function, but uses a more performant approach for generating immutable arrays as it reduces the number of necessary array copies.
 *
 * `applyCombinations([], [[1, 2], [3, 4]])` for example produces:
 * - `[1, 3]`
 * - `[1, 4]`
 * - `[2, 3]`
 * - `[2, 4]`
 *
 * If the values of `source` are equal to the values of one of the combinations, it is used instead of a copy in the output array, e.g.:
 * `let i = [1, 2]; applyCombinations(i, [[1], [2]])[0] == i`
 * - `true`
 *
 * @param {any[][]} combinations The combinations
 */
export function applyCombinations(source, combinations) {
  const l = combinations.length;
  let total = 1;
  const indices = combinations.map((entry) => {
    const count = entry.length;
    total *= count;
    return [count, 0, entry];
  });
  if (total === 0) return [];
  let s = source;
  const sl = source.length;
  if (sl !== l) s = sl > l ? s.slice(0, l) : [...s, ...new Array(l - sl)];
  const out = new Array(total);
  let c = 0;
  for (;;) {
    let result = s;
    let unchanged = true;
    indices.forEach(([, j, values], i) => {
      const v = values[j];
      if (result[i] === v) return;
      if (unchanged) {
        result = [...result];
        unchanged = false;
      }
      result[i] = v;
    });
    out[c] = result;
    c += 1;
    if (c >= total) break;
    for (let n = l - 1; n >= 0; n -= 1) {
      const i = indices[n];
      const [max, current] = i;
      const next = current + 1;
      if (next < max) {
        i[1] = next;
        break;
      }
      i[1] = 0;
    }
  }
  return out;
}

File: go/library/apply.go
package library

type Entry[Key any] struct {
	Key     Key
	NoValue bool
	Value   any
}
type ObjectEntry = Entry[string]
type ArrayEntry = Entry[int]

// TODO: this will not work like expected
// because for example slices cannot be compared
// and are not passed by pointer but are copied.
//
// One possible workaround would be a deep comparison
// but this would be very imperformant,
// the other workaround would be to pass arrays and
// objects as slice / map pointers instead of them directly,
// however that would make the transition to
// other JSON packages rather complicated and
// expensive in terms of memory.
//
// A third (and maybe best) workaround would be
// to use unsafe.SliceData
// as described here:
// https://stackoverflow.com/questions/53009686/equality-identity-of-go-slices
// However, there still probably needs to be found a way to compare maps (if they cannot simply be compared directly).
// UPDATE:
// reflect.ValueOf(a).Pointer() == reflect.ValueOf(b).Pointer()
// works for both slices and maps

// Apply all changes immutably to the source object.
func applyObject(source map[string]any, changes []ObjectEntry) map[string]any {
	result := source
	unchanged := true

	for _, change := range changes {
		if change.NoValue {
			if _, exists := result[change.Key]; !exists {
				continue
			}
			if unchanged {
				result = copyMap(result)
				delete(result, change.Key)
				unchanged = false
			} else {
				delete(result, change.Key)
			}
			continue
		}

		if value, exists := result[change.Key]; exists && value == change.Value {
			continue
		}
		if unchanged {
			result = copyMap(result)
			result[change.Key] = change.Value
			unchanged = false
		} else {
			result[change.Key] = change.Value
		}
	}

	return result
}

// Apply all changes immutably to the source array.
// Indices can be negative to be applied from the end of the array.
func applyArray(source []any, changes []ArrayEntry, filler any) []any {
	result := source
	unchanged := true

	for _, change := range changes {
		i := change.Key
		if change.Key < 0 {
			i = len(result) + change.Key
		}

		if i >= 0 && i < len(result) && result[i] == change.Value {
			continue
		}
		if change.Key >= 0 {
			suf := i + 1 - len(result)
			if suf > 0 {
				total := len(result) + suf
				nextResult := make([]any, total)
				for i := copy(nextResult, result); i < total; i++ {
					nextResult[i] = filler
				}
				result = nextResult
				unchanged = false
			}
		} else {
			pre := -i
			if pre > 0 {
				total := pre + len(result)
				nextResult := make([]any, total)
				for i := 0; i < pre; i++ {
					nextResult[i] = filler
				}
				copy(nextResult[pre:], result)
				result = nextResult
				unchanged = false
				i = 0
			}
		}
		if unchanged {
			nextResult := make([]any, len(result))
			copy(nextResult, result)
			result = nextResult
			unchanged = false
		}
		result[i] = change.Value
	}

	return result
}

type combinationIndex struct {
	count   int
	current int
	values  []any
}

// Create all possible combinations immutably.
// If the specified `combinations` array is empty, the resulting array contains a single empty array.
// This function has essentially the same base functionality as the `mux` function, but uses a more performant approach for generating immutable arrays as it reduces the number of necessary array copies.
//
// `applyCombinations([], [[1, 2], [3, 4]])` for example produces:
// - `[1, 3]`
// - `[1, 4]`
// - `[2, 3]`
// - `[2, 4]`
func applyCombinations(combinations [][]any) []any {
	l := len(combinations)
	total := 1
	indices := make([]*combinationIndex, l)
	for i, entry := range combinations {
		count := len(entry)
		total *= count
		indices[i] = &combinationIndex{count, 0, entry}
	}
	if total == 0 {
		return nil
	}
	s := make([]any, l)
	out := make([]any, total)
	var c int
	for {
		result := s
		unchanged := true
		for i, index := range indices {
			v := index.values[index.current]
			if result[i] == v {
				return
			}
		}
	}
	return out
}

// Copies the source map
func copyMap[Value any](source map[string]Value) (result map[string]Value) {
	result = make(map[string]Value, len(source))
	for key, value := range source {
		result[key] = value
	}
	return
}
